modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java

```diff
@@ -265,19 +269,47 @@ protected int processStartTag() throws Exception {
 
 		jspWriter.write("</li>");
 
-		int totalPages = (int)Math.ceil(_totalItems / _activeDelta);
+		int totalPages = (int)Math.ceil(
+			(double)_totalItems / (double)_activeDelta);
+
+		if (_ellipsisBuffer != 0) {
+			int activeIndex = _activePage - 1;
+
+			int[] totalItems = new int[totalPages];
+
+			for (int i = 0; i < totalPages; i++) {
+				totalItems[i] = i + 1;
+			}
+
+			int lastIndex = totalItems.length - 1;
+			int leftBufferEnd = activeIndex - _ellipsisBuffer;
+			int rightBufferStart = activeIndex + _ellipsisBuffer + 1;
 
-		int ellipsisCount = _activePage + _ellipsisBuffer;
+			int[] leftBuffer = Arrays.copyOfRange(
+				totalItems, 1, Math.max(leftBufferEnd, 1));
+
+			if (rightBufferStart > lastIndex) {
+				rightBufferStart = lastIndex;
+			}
+
+			int[] rightBuffer = Arrays.copyOfRange(
+				totalItems, rightBufferStart,
+				Math.max(lastIndex, rightBufferStart));
+
+			int[] itemsToRender = Arrays.copyOfRange(
+				totalItems, Math.max(activeIndex - _ellipsisBuffer, 1),
+				Math.min(activeIndex + _ellipsisBuffer + 1, lastIndex));
+
+			int firstItem = totalItems[0];
```

Comment:
Could this be null?

Reply:
I don't think it would

Reply:
It shouldn't be null if that is what you mean, but if there's a cleaner way to do it, I'm always interested.

Reply:
I was meaning that, it there is a case where totalItems is null. If it was you could safe guard with an if or use an Optional.

Reply:
`totalItems` should be a mandatory property to render the minimo but I'm not sure if we add an initial value even when it's not passed. I imagine that because it is mandatory in taglib it must fail at compile time.

Reply:
`totalItems` is already a [required](https://github.com/liferay/liferay-portal/blob/7da6852989d3c87e1d763eca20bfde198c1a8afe/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld#L2221-L2226) attribute. In the React component, it looks like the "prop" [isn't optional](https://github.com/liferay/clay/blob/386008b990fc2670b3944b538de29531d49c6d59/packages/clay-pagination/src/PaginationWithBasicItems.tsx#L60) either.